### ISSUE TITLE
[new feature] capture information from training dynamics

### ISSUE THREAD

**Issue by jwmueller** [cleanlab/cleanlab#974]
Goal: create a new module (for now say: `cleanlab.experimental.training_dynamics`) that allows users to provide model outputs/info at every iteration (aka checkpoint) of an iteratively trained model (eg. neural network). 

Useful things to record at each checkpoint include:

- prediction loss for each training datapoint
- predicted probabilities for each training datapoint

The code to capture these should be a simple logger that users can easily integrate with arbitrary iterative ML models (huggingface, pytorch, jax, keras, xgboost, ...).

Once these values are captured, they can be used in various functions, such as ones implementing the methods from these papers:

[TRIAGE: Characterizing and auditing training data for improved regression](https://arxiv.org/abs/2310.18970)

[Learning from Training Dynamics: Identifying Mislabeled Data beyond Manually Designed Features](https://ojs.aaai.org/index.php/AAAI/article/view/25972)

[Identifying Mislabeled Data using the Area Under the Margin Ranking](https://proceedings.neurips.cc/paper_files/paper/2020/file/c6102b3727b2a7d8b1bb6981147081ef-Paper.pdf)

**Comment by tataganesh** [comment]
Hey @jwmueller,
Are you still looking for contributions for this task?

**Comment by jwmueller** [comment]
> Hey @jwmueller, Are you still looking for contributions for this task?

Yes! Just be aware this might be a major undertaking, but feel free to take on the challenge! 

Other easier areas where we could really use your help include:

1. Any remaining improvement opportunities you spot for the underperforming group detection and associated clustering algorithm, or our documentation/tutorial for this.
2. Publishing a cleanlab.ai/blog post showcasing your underperforming group detection on an interesting dataset (ideally say a real text dataset without pre-defined slices, and a real tabular dataset with pre-defined slices).
3. Helping us review + merge this PR on identifier-column detection:
https://github.com/cleanlab/cleanlab/pull/1120
and pushing follow-up improvements to this new issue type.

Regardless excited to see any further contributions from you!

**Comment by tataganesh** [comment]
That makes sense, thank you! I am definitely more familiar with the underperforming group detection, so I can try and work on the blog/improvements for it. Will try and squeeze in the PR review as well once I have some more bandwidth!

**Comment by giovannicozzolongo** [comment]
Hi, I can start with the logger for per-sample losses and probabilities, designed to work with different frameworks. Could you assign the issue to me?
